src/openzeppelin/token/erc1155/library.cairo

```diff
@@ -0,0 +1,600 @@
+end
+
+@storage_var
+func ERC1155_operator_approvals(account : felt, operator : felt) -> (approved : felt):
+end
+
+@storage_var
+func ERC1155_uri() -> (uri : felt):
+end
+
+namespace ERC1155:
+    #
+    # Constructor
+    #
+
+    func initializer{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        uri_ : felt
+    ):
+        _setURI(uri_)
+        ERC165.register_interface(IERC1155_ID)
+        ERC165.register_interface(IERC1155_METADATA_ID)
+        return ()
+    end
+
+    #
+    # Getters
+    #
+
+    func uri{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}() -> (
+        uri : felt
+    ):
+        let (uri) = ERC1155_uri.read()
+        return (uri)
+    end
+
+    func balance_of{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        account : felt, id : Uint256
+    ) -> (balance : Uint256):
+        with_attr error_message("ERC1155: balance query for the zero address"):
+            assert_not_zero(account)
+        end
+        let (balance) = ERC1155_balances.read(id=id, account=account)
+        return (balance)
+    end
+
+    func balance_of_batch{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        accounts_len : felt, accounts : felt*, ids_len : felt, ids : Uint256*
+    ) -> (batch_balances_len : felt, batch_balances : Uint256*):
+        alloc_locals
+        # Check args are equal length arrays
+        with_attr error_message("ERC1155: accounts and ids length mismatch"):
+            assert ids_len = accounts_len
+        end
+        # Allocate memory
+        let (local batch_balances : Uint256*) = alloc()
+        let len = accounts_len
+        # Call iterator
+        balance_of_batch_iter(len, accounts, ids, batch_balances)
+        return (batch_balances_len=len, batch_balances=batch_balances)
+    end
+
+    func is_approved_for_all{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        account : felt, operator : felt
+    ) -> (approved : felt):
+        let (approved) = ERC1155_operator_approvals.read(account=account, operator=operator)
+        return (approved)
+    end
+
+    #
+    # Externals
+    #
+
+    func set_approval_for_all{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        operator : felt, approved : felt
+    ):
+        let (caller) = get_caller_address()
+        with_attr error_message("ERC1155: called from zero address"):
+            assert_not_zero(caller)
+        end
+        _set_approval_for_all(owner=caller, operator=operator, approved=approved)
+        return ()
+    end
+
+    func safe_transfer_from{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        from_ : felt, to : felt, id : Uint256, amount : Uint256, data_len : felt, data : felt*
+    ):
+        let (caller) = get_caller_address()
+        with_attr error_message("ERC1155: called from zero address"):
```

Comment:
I think the rationale here is that you "could" call contract from the 0 address as an approved operator. In which case 0 would be the caller, but not necessarily the from field.  Although since this shouldn't be possible due the the guards on approving the 0 address, this shouldn't occur. Is it worth distinguishing or even protecting against this (since owner_or_approved should already handle this)?

Reply:
Yeah, I think that error message works. And instead of explicitly adding another check, what do you think about leveraging the zero address check in `balance_of` and inserting that into [line 204](https://github.com/dewi-tim/cairo-contracts/blob/feature/erc1155-%23118/src/openzeppelin/token/erc1155/library.cairo#L204) to fetch the balance?